chisel/src/driver.rs

```diff
@@ -0,0 +1,228 @@
+                        DriverError::Internal(
+                            name.clone(),
+                            "Deserialization failure".to_string(),
+                            e.into(),
+                        ),
+                        results,
+                    );
+                    return &self.state;
+                }
+            };
+
+            // Consume modules in ruleset and execute.
+            while let Some((name, module)) = ruleset.modules_mut().pop_front() {
+                chisel_debug!(1, "Executing module {}", &name);
+
+                // TODO: homogeneous module handling
+                let module_result = match name.as_str() {
+                    _ => unimplemented!(),
+                };
+
+                // If the module was a translator or creator, we set the output in the result.
+                match module_result {
+                    ModuleResult::Creator(_, ref result)
+                    | ModuleResult::Translator(_, ref result) => {
+                        if let Ok(true) = result {
+                            chisel_debug!(1, "Module mutated or created.");
+                            ruleset_result.set_output_module(wasm.clone()); //TODO: Refactor to only set this at the end and save some expensive copies
+                        }
+                    }
+                    ModuleResult::Validator(_, _) => (),
+                }
+                ruleset_result.results_mut().push(module_result);
+            }
+            results.rulesets_mut().push(ruleset_result);
+        }
+        self.state = DriverState::Done(results);
+        &self.state
+    }
+}
+
+// Error.description() is deprecated for displaying errors now.
+impl Display for DriverError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            DriverError::MissingRequiredField(object, field) => {
+                write!(f, "in '{}': missing required field '{}'", object, field)
+            }
+            DriverError::ModuleNotFound(object, module) => {
+                write!(f, "in '{}': module '{}' not found", object, module)
+            }
+            DriverError::InvalidField(object, field) => {
+                write!(f, "in '{}': invalid field '{}'", object, field)
+            }
+            DriverError::PathResolution(object, path) => {
+                write!(f, "in '{}': failed to resolve path '{}'", object, path)
+            }
+            DriverError::Internal(object, info, err) => {
+                write!(f, "in '{}': {}; {}", object, info, err)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::panic::catch_unwind;
+
+    use super::*;
+    use crate::config::{ChiselConfig, FromArgs};
+
+    #[test]
+    fn take_result_ready() {
+        let result = catch_unwind(|| {
+            let config = ChiselConfig::from_args("test", "test.foo=bar").expect("Cannot fail");
+            let driver = ChiselDriver::new(config);
```

Comment:
Can you move these two out so that unwind only catches the panic in `take_result`?

Reply:
No, it doesn't allow config and driver to cross unwind boundaries.

Reply:
Or add another test case with these two lines succeeding.

Reply:
Even with `unsafe`

Reply:
The success test requires fire().